File: Data Structures & Algorithms/three-integer-sum/submission-4.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        List<List<Integer>> list = new ArrayList<List<Integer>>();
        Arrays.sort(nums);
        //test against prefex

        //prex ---- nums
        for(int i = 0; i <= nums.length - 3; i++){
            //posive check
            if(nums[i] > 0) break; 
            //skip the prefix thats the same
            if (i > 0 && nums[i] == nums[i-1]) continue;

            int j = i + 1;
            int k = nums.length - 1;
            while(j < k){
                int sum = nums[i] + nums[j] + nums[k]; 
                if(sum == 0){
                    list.add(Arrays.asList(nums[i], nums[j], nums[k]));
                    j++;
                    k--;

                    //skip dupes
                    while(j < k && nums[j] == nums[j - 1]){
                        j++;
                    } 

                    while(j < k && nums[k] == nums[k + 1]){
                        k--;
                    }
                }else if (sum > 0){
                    k--;
                }else{
                    j++;
                }
                
            }

        }


        return list;
    }
}
//i , j, k = 0;

File: Data Structures & Algorithms/sqrtx/submission-4.cs
public class Solution {
    public int MySqrt(int x) {
        int l = 1;
        int r = x; 
        int res = 0;
        while(l <= r){
            var mid = l + (r - l ) / 2;

            long sum = (long)mid * mid;
            if(sum == x){
                return mid;
            }else if (sum < x){
                l = mid + 1;
                res = mid;
            }else{
                r = mid - 1;
            }
        }

        return res;
    }
}